ddl/db_test.go

```diff
@@ -4833,6 +4833,28 @@ func (s *testDBSuite1) TestAlterTableWithValidation(c *C) {
 	tk.MustQuery("show warnings").Check(testutil.RowsWithSep("|", "Warning|8200|ALTER TABLE WITHOUT VALIDATION is currently unsupported"))
 }
 
+func (s *testDBSuite1) TestAlterTableAlterPartition(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t1")
+	defer tk.MustExec("drop table if exists t1")
+
+	tk.MustExec(`create table t (c int)
+PARTITION BY RANGE (c) (
+	PARTITION p0 VALUES LESS THAN (6),
+	PARTITION p1 VALUES LESS THAN (11),
+	PARTITION p2 VALUES LESS THAN (16),
+	PARTITION p3 VALUES LESS THAN (21)
+);`)
+
+	_, err := tk.Exec(`alter table t alter partition p0
+add placement policy
+	constraints='+zone=sh'
+	role=leader
+	replicas=3`)
+	c.Assert(err, ErrorMatches, ".*pd unavailable.*")
```

Comment:
Please add some exceptional cases, like:
1. The table is actually not partitioned
2. Adding rules to a non-existed partition

Besides, please add various `ALTER TABLE` statements to test.

Reply:
`ALTER TABLE t ADD PLACEMNT POLICY`? That is not supported yet. Other things it sounds good to me.

Reply:
https://github.com/pingcap/tidb/pull/18611/commits/72a5e99e904361f2d5da61f30c8eb6d0a1d77689 adds two more tests. Some more tests are added in the validation commit.